sqlite_data_norm.c

```diff
@@ -0,0 +1,303 @@
+ *	if (pBlob == 0)
+ *		return;
+ *	sqlite_fdw_data_norm_uuid_blob_to_str(pBlob, zs);
+ *	sqlite3_result_text(context, (char*)zs, 36, SQLITE_TRANSIENT);
+ *}
+ */
+
+/*
+ * uuid_blob normalize text or blob UUID argv[0] into a 16-byte blob.
+ */
+static void
+sqlite_fdw_data_norm_uuid(sqlite3_context* context, int argc, sqlite3_value** argv)
+{
+	unsigned char aBlob[16];
+	sqlite3_value* arg = argv[0];		
+	
+	if (sqlite3_value_type(argv[0]) == SQLITE3_TEXT)
+	{
+		const unsigned char* txt = sqlite3_value_text(arg);
+		if (sqlite_fdw_uuid_blob(txt, aBlob))
+		{
+			sqlite3_result_blob(context, aBlob, 16, SQLITE_TRANSIENT);
+			return;
+		}
+	}
+	sqlite3_result_value(context, arg);
+}
+
+/*
+ * ISO:SQL valid boolean values with text affinity such as Y, no, f, t, oN etc.
+ * will be treated as boolean like in PostgreSQL console input
+ */ 
+static void
+sqlite_fdw_data_norm_bool(sqlite3_context* context, int argc, sqlite3_value** argv)
+{
+
+	sqlite3_value* arg = argv[0];
+	int dt = sqlite3_value_type(arg);
+	const char* t;
+	int l;
+
+	if (dt == SQLITE_INTEGER)
+	{
+		/* The fastest call because expected very often */
+		sqlite3_result_value(context, arg);
+		return;
+	}
+	if (dt != SQLITE3_TEXT && dt != SQLITE_BLOB )
+	{
+		/* NULL, FLOAT */
+		sqlite3_result_value(context, arg);
+		return;
+	}
+	l = sqlite3_value_bytes(arg);
+	if (l > 5)
+	{
+		sqlite3_result_value(context, arg);
+		return;
+	}
+	
+	t = (const char*)sqlite3_value_text(arg);
+		
+	if ( l == 1 )
+	{
+		if (strcasecmp(t, "t") == 0)
+		{
+			sqlite3_result_int(context, 1);
+			return;
+		}
+		if (strcasecmp(t, "f") == 0)
+		{
+			sqlite3_result_int(context, 0);
+			return;
+		}
+		if (strcasecmp(t, "y") == 0)
+		{
+			sqlite3_result_int(context, 1);
+			return;
+		}
+		if (strcasecmp(t, "n") == 0)
+		{
+			sqlite3_result_int(context, 0);
+			return;
+		}
+		/* rare but possible cases */
+		if (strcasecmp(t, "1") == 0)
+		{
+			sqlite3_result_int(context, 1);
+			return;
+		}
+		if (strcasecmp(t, "0") == 0)
+		{
+			sqlite3_result_int(context, 0);
+			return;
+		}
+	}
+	if ( l == 2 )
+	{			
+		if (strcasecmp(t, "on") == 0)
+		{
+			sqlite3_result_int(context, 1);
+			return;
+		}
+		if (strcasecmp(t, "no") == 0)
+		{
+			sqlite3_result_int(context, 0);
+			return;
+		}
+	}
+	if ( l == 3 )
+	{
+		if (strcasecmp(t, "yes") == 0)
+		{
+			sqlite3_result_int(context, 1);
+			return;
+		}
+		if (strcasecmp(t, "off") == 0)
+		{
+			sqlite3_result_int(context, 0);
+			return;
+		}
+	}
+	if ( l == 4 && strcasecmp(t, "true") == 0)
+	{
+		sqlite3_result_int(context, 1);
+		return;
+	}
+	if ( l == 5 && strcasecmp(t, "false") == 0)
+	{
+		sqlite3_result_int(context, 0);
+		return;
+	}
+	sqlite3_result_value(context, arg);
+}
+
+int
+sqlite_fdw_data_norm_functs_init(sqlite3* db) {
+	
+	static const int det_flags = SQLITE_UTF8 | SQLITE_INNOCUOUS | SQLITE_DETERMINISTIC;
+
+	int rc = sqlite3_create_function(db, "sqlite_fdw_uuid_blob", 1, det_flags, 0, sqlite_fdw_data_norm_uuid, 0, 0);	
```

Comment:
Please check the return code and raise an error by ereport()/elog() with an error message acquired by sqlites_errmsg() if an error occurred.

The return type of function will be `void`. So could you fix connection.c also where calling sqlite_fdw_data_norm_functs_init()?

Reply:
>  ... could you fix connection.c also where calling sqlite_fdw_data_norm_functs_init()?

Yes. Distributed error processing is bad practice for this very simple case. Concentrated in `sqlite_data_norm.c` in https://github.com/pgspider/sqlite_fdw/commit/72e1b22177db37251b2cda4474eb46955d1d6f26. Please recheck, @t-kataym .

Reply:
Thank you for fixing.